# Dataset - Hypothyroid

In [1]:
import pandas as pd
import numpy as np

In [2]:
hthyroid=pd.read_csv('dataset_57_hypothyroid.csv',sep=",")

In [3]:
hthyroid.head()

age sex on_thyroxine query_on_thyroxine on_antithyroid_medication sick  \
0  41   F            f                  f                         f    f   
1  23   F            f                  f                         f    f   
2  46   M            f                  f                         f    f   
3  70   F            t                  f                         f    f   
4  70   F            f                  f                         f    f   

  pregnant thyroid_surgery I131_treatment query_hypothyroid  ... TT4_measured  \
0        f               f              f                 f  ...            t   
1        f               f              f                 f  ...            t   
2        f               f              f                 f  ...            t   
3        f               f              f                 f  ...            t   
4        f               f              f                 f  ...            t   

   TT4 T4U_measured   T4U FTI_measured  FTI TBG_measured TBG referral_source  \
0  125            t  1.14            t  109            f   ?            SVHC   
1  102            f     ?            f    ?            f   ?           other   
2  109            t  0.91            t  120            f   ?           other   
3  175            f     ?            f    ?            f   ?           other   
4   61            t  0.87            t   70            f   ?             SVI   

      Class  
0  negative  
1  negative  
2  negative  
3  negative  
4  negative  

[5 rows x 30 columns]

In [4]:
hthyroid.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3772 entries, 0 to 3771
Data columns (total 30 columns):
age                          3772 non-null object
sex                          3772 non-null object
on_thyroxine                 3772 non-null object
query_on_thyroxine           3772 non-null object
on_antithyroid_medication    3772 non-null object
sick                         3772 non-null object
pregnant                     3772 non-null object
thyroid_surgery              3772 non-null object
I131_treatment               3772 non-null object
query_hypothyroid            3772 non-null object
query_hyperthyroid           3772 non-null object
lithium                      3772 non-null object
goitre                       3772 non-null object
tumor                        3772 non-null object
hypopituitary                3772 non-null object
psych                        3772 non-null object
TSH_measured                 3772 non-null object
TSH                          3772 non-null 

In [5]:
data=hthyroid.copy()

In [6]:
data=data.replace('?',np.nan)

In [7]:
data.head()

age sex on_thyroxine query_on_thyroxine on_antithyroid_medication sick  \
0  41   F            f                  f                         f    f   
1  23   F            f                  f                         f    f   
2  46   M            f                  f                         f    f   
3  70   F            t                  f                         f    f   
4  70   F            f                  f                         f    f   

  pregnant thyroid_surgery I131_treatment query_hypothyroid  ... TT4_measured  \
0        f               f              f                 f  ...            t   
1        f               f              f                 f  ...            t   
2        f               f              f                 f  ...            t   
3        f               f              f                 f  ...            t   
4        f               f              f                 f  ...            t   

   TT4 T4U_measured   T4U FTI_measured  FTI TBG_measured TBG referral_source  \
0  125            t  1.14            t  109            f NaN            SVHC   
1  102            f   NaN            f  NaN            f NaN           other   
2  109            t  0.91            t  120            f NaN           other   
3  175            f   NaN            f  NaN            f NaN           other   
4   61            t  0.87            t   70            f NaN             SVI   

      Class  
0  negative  
1  negative  
2  negative  
3  negative  
4  negative  

[5 rows x 30 columns]

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3772 entries, 0 to 3771
Data columns (total 30 columns):
age                          3771 non-null object
sex                          3622 non-null object
on_thyroxine                 3772 non-null object
query_on_thyroxine           3772 non-null object
on_antithyroid_medication    3772 non-null object
sick                         3772 non-null object
pregnant                     3772 non-null object
thyroid_surgery              3772 non-null object
I131_treatment               3772 non-null object
query_hypothyroid            3772 non-null object
query_hyperthyroid           3772 non-null object
lithium                      3772 non-null object
goitre                       3772 non-null object
tumor                        3772 non-null object
hypopituitary                3772 non-null object
psych                        3772 non-null object
TSH_measured                 3772 non-null object
TSH                          3403 non-null 

In [9]:
data.isnull().sum()

age                             1
sex                           150
on_thyroxine                    0
query_on_thyroxine              0
on_antithyroid_medication       0
sick                            0
pregnant                        0
thyroid_surgery                 0
I131_treatment                  0
query_hypothyroid               0
query_hyperthyroid              0
lithium                         0
goitre                          0
tumor                           0
hypopituitary                   0
psych                           0
TSH_measured                    0
TSH                           369
T3_measured                     0
T3                            769
TT4_measured                    0
TT4                           231
T4U_measured                    0
T4U                           387
FTI_measured                    0
FTI                           385
TBG_measured                    0
TBG                          3772
referral_source                 0
Class         

In [10]:
data.loc[data.TSH_measured=='f']['TSH'].isnull().sum()

369

In [11]:
data.loc[data.T3_measured=='f']['T3'].isnull().sum()

769

In [12]:
data.loc[data.TT4_measured=='f']['TT4'].isnull().sum()

231

In [13]:
data.loc[data.T4U_measured=='f']['T4U'].isnull().sum()

387

In [14]:
data.loc[data.FTI_measured=='f']['FTI'].isnull().sum()

385

It is seen that when the categorical variables (as seen above) are false, then there are missing values for the numerical data in the respective columns. Hence we drop these numerical columns. 

In [15]:
data=data.drop(columns=['TSH','T3','TT4','T4U','FTI','TBG'])

In [16]:
data.sample(5)

age sex on_thyroxine query_on_thyroxine on_antithyroid_medication sick  \
1184  42   F            f                  f                         f    f   
1477  34   F            t                  f                         f    f   
145   37   F            f                  f                         f    f   
2229  73   F            f                  f                         f    f   
1751  42   F            f                  f                         f    f   

     pregnant thyroid_surgery I131_treatment query_hypothyroid  ...  \
1184        f               f              f                 f  ...   
1477        t               f              f                 f  ...   
145         f               f              f                 f  ...   
2229        f               f              f                 f  ...   
1751        t               f              f                 f  ...   

     hypopituitary psych TSH_measured T3_measured TT4_measured T4U_measured  \
1184             f     f            t           t            t            t   
1477             f     f            t           t            t            t   
145              f     f            t           t            t            t   
2229             f     f            t           t            t            t   
1751             f     f            t           t            t            t   

     FTI_measured TBG_measured referral_source     Class  
1184            t            f           other  negative  
1477            t            f           other  negative  
145             t            f            SVHC  negative  
2229            t            f           other  negative  
1751            t            f            STMW  negative  

[5 rows x 24 columns]

In [17]:
data.isnull().sum()

age                            1
sex                          150
on_thyroxine                   0
query_on_thyroxine             0
on_antithyroid_medication      0
sick                           0
pregnant                       0
thyroid_surgery                0
I131_treatment                 0
query_hypothyroid              0
query_hyperthyroid             0
lithium                        0
goitre                         0
tumor                          0
hypopituitary                  0
psych                          0
TSH_measured                   0
T3_measured                    0
TT4_measured                   0
T4U_measured                   0
FTI_measured                   0
TBG_measured                   0
referral_source                0
Class                          0
dtype: int64

In [18]:
data[data.pregnant=='t']['sex'].isnull().sum()

2

So the sex for these datasets can be replced with 'F'

In [19]:
z=data[(data.pregnant=='t') & (data.sex.isnull())].index

In [20]:
data.loc[z,'sex']=data.loc[z,'sex'].fillna('F')

In [21]:
data.loc[z]

age sex on_thyroxine query_on_thyroxine on_antithyroid_medication sick  \
1609  73   F            f                  f                         f    f   
3271  21   F            f                  f                         f    f   

     pregnant thyroid_surgery I131_treatment query_hypothyroid  ...  \
1609        t               f              f                 f  ...   
3271        t               f              f                 f  ...   

     hypopituitary psych TSH_measured T3_measured TT4_measured T4U_measured  \
1609             f     f            t           t            t            t   
3271             f     f            t           t            t            t   

     FTI_measured TBG_measured referral_source     Class  
1609            t            f           other  negative  
3271            t            f            STMW  negative  

[2 rows x 24 columns]

In [22]:
data.isnull().sum()

age                            1
sex                          148
on_thyroxine                   0
query_on_thyroxine             0
on_antithyroid_medication      0
sick                           0
pregnant                       0
thyroid_surgery                0
I131_treatment                 0
query_hypothyroid              0
query_hyperthyroid             0
lithium                        0
goitre                         0
tumor                          0
hypopituitary                  0
psych                          0
TSH_measured                   0
T3_measured                    0
TT4_measured                   0
T4U_measured                   0
FTI_measured                   0
TBG_measured                   0
referral_source                0
Class                          0
dtype: int64

In [23]:
data['age']=pd.to_numeric(data.age,errors='coerce') #converting it to a numerical column

In [24]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3772 entries, 0 to 3771
Data columns (total 24 columns):
age                          3771 non-null float64
sex                          3624 non-null object
on_thyroxine                 3772 non-null object
query_on_thyroxine           3772 non-null object
on_antithyroid_medication    3772 non-null object
sick                         3772 non-null object
pregnant                     3772 non-null object
thyroid_surgery              3772 non-null object
I131_treatment               3772 non-null object
query_hypothyroid            3772 non-null object
query_hyperthyroid           3772 non-null object
lithium                      3772 non-null object
goitre                       3772 non-null object
tumor                        3772 non-null object
hypopituitary                3772 non-null object
psych                        3772 non-null object
TSH_measured                 3772 non-null object
T3_measured                  3772 non-null

In [25]:
data['age'].describe()

count    3771.000000
mean       51.735879
std        20.084958
min         1.000000
25%        36.000000
50%        54.000000
75%        67.000000
max       455.000000
Name: age, dtype: float64

We can see that the age column ahs outliers (min=1, max=455). Hence we first find the outliers

In [26]:
data[data.age>90]

age sex on_thyroxine query_on_thyroxine on_antithyroid_medication  \
1129   94.0   F            f                  f                         f   
1364  455.0   F            f                  f                         f   
1490   91.0   M            f                  f                         f   
1727   91.0   F            f                  f                         f   
2418   93.0   F            f                  f                         f   
2673   94.0   M            f                  f                         f   
2760   92.0   F            t                  f                         f   
2804   92.0   F            f                  f                         f   
3014   93.0   F            f                  f                         f   

     sick pregnant thyroid_surgery I131_treatment query_hypothyroid  ...  \
1129    f        f               f              f                 f  ...   
1364    f        f               f              f                 f  ...   
1490    f        f               f              f                 f  ...   
1727    f        f               f              f                 f  ...   
2418    t        f               f              f                 f  ...   
2673    f        f               f              f                 f  ...   
2760    f        f               f              f                 f  ...   
2804    f        f               f              f                 f  ...   
3014    t        f               f              f                 f  ...   

     hypopituitary psych TSH_measured T3_measured TT4_measured T4U_measured  \
1129             f     f            t           f            t            t   
1364             f     f            t           t            t            t   
1490             f     f            t           f            t            t   
1727             f     f            t           t            t            t   
2418             f     f            t           t            t            t   
2673             f     f            t           t            t            t   
2760             f     f            t           f            f            f   
2804             f     f            t           t            t            t   
3014             f     f            t           t            t            t   

     FTI_measured TBG_measured referral_source     Class  
1129            t            f             SVI  negative  
1364            t            f             SVI  negative  
1490            t            f             SVI  negative  
1727            t            f             SVI  negative  
2418            t            f           other  negative  
2673            t            f             SVI  negative  
2760            f            f           other  negative  
2804            t            f             SVI  negative  
3014            t            f             SVI  negative  

[9 rows x 24 columns]

Removing the row which has the age as 455

In [27]:
data=data.drop(index=1364)

Even infants can suffer from hypothyroidism. Hence we are not eliminating the records with lower values of age

In [28]:
data['age'].describe()

count    3770.000000
mean       51.628912
std        18.982990
min         1.000000
25%        36.000000
50%        54.000000
75%        67.000000
max        94.000000
Name: age, dtype: float64

Only 4% of the data is lost if we drop the rows with any missing values.

In [29]:
data=data.dropna(axis=0,how='any') # drrop all the rows which have any missing values

In [30]:
data.isnull().sum()

age                          0
sex                          0
on_thyroxine                 0
query_on_thyroxine           0
on_antithyroid_medication    0
sick                         0
pregnant                     0
thyroid_surgery              0
I131_treatment               0
query_hypothyroid            0
query_hyperthyroid           0
lithium                      0
goitre                       0
tumor                        0
hypopituitary                0
psych                        0
TSH_measured                 0
T3_measured                  0
TT4_measured                 0
T4U_measured                 0
FTI_measured                 0
TBG_measured                 0
referral_source              0
Class                        0
dtype: int64

In [31]:
for i in data.columns[1:]:
    print(data[i].value_counts())
    print('\n')

F    2480
M    1142
Name: sex, dtype: int64


f    3174
t     448
Name: on_thyroxine, dtype: int64


f    3573
t      49
Name: query_on_thyroxine, dtype: int64


f    3579
t      43
Name: on_antithyroid_medication, dtype: int64


f    3479
t     143
Name: sick, dtype: int64


f    3569
t      53
Name: pregnant, dtype: int64


f    3570
t      52
Name: thyroid_surgery, dtype: int64


f    3568
t      54
Name: I131_treatment, dtype: int64


f    3393
t     229
Name: query_hypothyroid, dtype: int64


f    3393
t     229
Name: query_hyperthyroid, dtype: int64


f    3604
t      18
Name: lithium, dtype: int64


f    3588
t      34
Name: goitre, dtype: int64


f    3527
t      95
Name: tumor, dtype: int64


f    3621
t       1
Name: hypopituitary, dtype: int64


f    3443
t     179
Name: psych, dtype: int64


t    3270
f     352
Name: TSH_measured, dtype: int64


t    2877
f     745
Name: T3_measured, dtype: int64


t    3405
f     217
Name: TT4_measured, dtype: int64


t    3255
f     367
N

TBG_measured column has only one variable throughout. Hence it will not be useful for us.

In [32]:
data=data.drop(columns='TBG_measured')

In [33]:
data.head()

age sex on_thyroxine query_on_thyroxine on_antithyroid_medication sick  \
0  41.0   F            f                  f                         f    f   
1  23.0   F            f                  f                         f    f   
2  46.0   M            f                  f                         f    f   
3  70.0   F            t                  f                         f    f   
4  70.0   F            f                  f                         f    f   

  pregnant thyroid_surgery I131_treatment query_hypothyroid  ... tumor  \
0        f               f              f                 f  ...     f   
1        f               f              f                 f  ...     f   
2        f               f              f                 f  ...     f   
3        f               f              f                 f  ...     f   
4        f               f              f                 f  ...     f   

  hypopituitary psych TSH_measured T3_measured TT4_measured T4U_measured  \
0             f     f            t           t            t            t   
1             f     f            t           t            t            f   
2             f     f            t           f            t            t   
3             f     f            t           t            t            f   
4             f     f            t           t            t            t   

  FTI_measured referral_source     Class  
0            t            SVHC  negative  
1            f           other  negative  
2            t           other  negative  
3            f           other  negative  
4            t             SVI  negative  

[5 rows x 23 columns]

In [34]:
data.isnull().sum()

age                          0
sex                          0
on_thyroxine                 0
query_on_thyroxine           0
on_antithyroid_medication    0
sick                         0
pregnant                     0
thyroid_surgery              0
I131_treatment               0
query_hypothyroid            0
query_hyperthyroid           0
lithium                      0
goitre                       0
tumor                        0
hypopituitary                0
psych                        0
TSH_measured                 0
T3_measured                  0
TT4_measured                 0
T4U_measured                 0
FTI_measured                 0
referral_source              0
Class                        0
dtype: int64

In [35]:
data.shape

(3622, 23)

In [36]:
data=data.reset_index(drop=True) # resetting the index to avoid missed rows

In [37]:
data.index

RangeIndex(start=0, stop=3622, step=1)

In [38]:
data.head()

age sex on_thyroxine query_on_thyroxine on_antithyroid_medication sick  \
0  41.0   F            f                  f                         f    f   
1  23.0   F            f                  f                         f    f   
2  46.0   M            f                  f                         f    f   
3  70.0   F            t                  f                         f    f   
4  70.0   F            f                  f                         f    f   

  pregnant thyroid_surgery I131_treatment query_hypothyroid  ... tumor  \
0        f               f              f                 f  ...     f   
1        f               f              f                 f  ...     f   
2        f               f              f                 f  ...     f   
3        f               f              f                 f  ...     f   
4        f               f              f                 f  ...     f   

  hypopituitary psych TSH_measured T3_measured TT4_measured T4U_measured  \
0             f     f            t           t            t            t   
1             f     f            t           t            t            f   
2             f     f            t           f            t            t   
3             f     f            t           t            t            f   
4             f     f            t           t            t            t   

  FTI_measured referral_source     Class  
0            t            SVHC  negative  
1            f           other  negative  
2            t           other  negative  
3            f           other  negative  
4            t             SVI  negative  

[5 rows x 23 columns]

### Normalization

In [39]:
import sklearn

In [40]:
from sklearn import preprocessing

In [41]:
x=data[['age']].values

In [42]:
x

array([[41.],
       [23.],
       [46.],
       ...,
       [74.],
       [72.],
       [64.]])

In [43]:
min_max_scaler = preprocessing.MinMaxScaler()

In [44]:
x_scaled = min_max_scaler.fit_transform(x)

In [45]:
df = pd.DataFrame(x_scaled)

In [46]:
df.head()

0
0  0.430108
1  0.236559
2  0.483871
3  0.741935
4  0.741935

In [47]:
df.columns=['Age_normalized']

In [48]:
df.shape

(3622, 1)

In [49]:
df.head()

Age_normalized
0        0.430108
1        0.236559
2        0.483871
3        0.741935
4        0.741935

In [50]:
df.isnull().sum()

Age_normalized    0
dtype: int64

### Label Encoding

In [51]:
enc=data.copy()

In [52]:
enc.columns

Index(['age', 'sex', 'on_thyroxine', 'query_on_thyroxine',
       'on_antithyroid_medication', 'sick', 'pregnant', 'thyroid_surgery',
       'I131_treatment', 'query_hypothyroid', 'query_hyperthyroid', 'lithium',
       'goitre', 'tumor', 'hypopituitary', 'psych', 'TSH_measured',
       'T3_measured', 'TT4_measured', 'T4U_measured', 'FTI_measured',
       'referral_source', 'Class'],
      dtype='object')

In [53]:
enc.head()

age sex on_thyroxine query_on_thyroxine on_antithyroid_medication sick  \
0  41.0   F            f                  f                         f    f   
1  23.0   F            f                  f                         f    f   
2  46.0   M            f                  f                         f    f   
3  70.0   F            t                  f                         f    f   
4  70.0   F            f                  f                         f    f   

  pregnant thyroid_surgery I131_treatment query_hypothyroid  ... tumor  \
0        f               f              f                 f  ...     f   
1        f               f              f                 f  ...     f   
2        f               f              f                 f  ...     f   
3        f               f              f                 f  ...     f   
4        f               f              f                 f  ...     f   

  hypopituitary psych TSH_measured T3_measured TT4_measured T4U_measured  \
0             f     f            t           t            t            t   
1             f     f            t           t            t            f   
2             f     f            t           f            t            t   
3             f     f            t           t            t            f   
4             f     f            t           t            t            t   

  FTI_measured referral_source     Class  
0            t            SVHC  negative  
1            f           other  negative  
2            t           other  negative  
3            f           other  negative  
4            t             SVI  negative  

[5 rows x 23 columns]

In [54]:
enc=enc.drop(columns=['age','Class'])

In [55]:
enc.shape

(3622, 21)

In [56]:
enc.index

RangeIndex(start=0, stop=3622, step=1)

In [57]:
from sklearn.preprocessing import LabelEncoder

In [58]:
le = LabelEncoder()

In [59]:
enc=enc.apply(le.fit_transform)

In [60]:
enc.head()

sex  on_thyroxine  query_on_thyroxine  on_antithyroid_medication  sick  \
0    0             0                   0                          0     0   
1    0             0                   0                          0     0   
2    1             0                   0                          0     0   
3    0             1                   0                          0     0   
4    0             0                   0                          0     0   

   pregnant  thyroid_surgery  I131_treatment  query_hypothyroid  \
0         0                0               0                  0   
1         0                0               0                  0   
2         0                0               0                  0   
3         0                0               0                  0   
4         0                0               0                  0   

   query_hyperthyroid  ...  goitre  tumor  hypopituitary  psych  TSH_measured  \
0                   0  ...       0      0              0      0             1   
1                   0  ...       0      0              0      0             1   
2                   0  ...       0      0              0      0             1   
3                   0  ...       0      0              0      0             1   
4                   0  ...       0      0              0      0             1   

   T3_measured  TT4_measured  T4U_measured  FTI_measured  referral_source  
0            1             1             1             1                1  
1            1             1             0             0                4  
2            0             1             1             1                4  
3            1             1             0             0                4  
4            1             1             1             1                3  

[5 rows x 21 columns]

In [61]:
enc['age']=df['Age_normalized']

In [62]:
enc['class']=data['Class']

In [63]:
enc.isnull().sum()

sex                          0
on_thyroxine                 0
query_on_thyroxine           0
on_antithyroid_medication    0
sick                         0
pregnant                     0
thyroid_surgery              0
I131_treatment               0
query_hypothyroid            0
query_hyperthyroid           0
lithium                      0
goitre                       0
tumor                        0
hypopituitary                0
psych                        0
TSH_measured                 0
T3_measured                  0
TT4_measured                 0
T4U_measured                 0
FTI_measured                 0
referral_source              0
age                          0
class                        0
dtype: int64

In [64]:
df_processed=enc.copy()

In [65]:
df_processed.head()

sex  on_thyroxine  query_on_thyroxine  on_antithyroid_medication  sick  \
0    0             0                   0                          0     0   
1    0             0                   0                          0     0   
2    1             0                   0                          0     0   
3    0             1                   0                          0     0   
4    0             0                   0                          0     0   

   pregnant  thyroid_surgery  I131_treatment  query_hypothyroid  \
0         0                0               0                  0   
1         0                0               0                  0   
2         0                0               0                  0   
3         0                0               0                  0   
4         0                0               0                  0   

   query_hyperthyroid  ...  hypopituitary  psych  TSH_measured  T3_measured  \
0                   0  ...              0      0             1            1   
1                   0  ...              0      0             1            1   
2                   0  ...              0      0             1            0   
3                   0  ...              0      0             1            1   
4                   0  ...              0      0             1            1   

   TT4_measured  T4U_measured  FTI_measured  referral_source       age  \
0             1             1             1                1  0.430108   
1             1             0             0                4  0.236559   
2             1             1             1                4  0.483871   
3             1             0             0                4  0.741935   
4             1             1             1                3  0.741935   

      class  
0  negative  
1  negative  
2  negative  
3  negative  
4  negative  

[5 rows x 23 columns]

In [66]:
df_processed.shape

(3622, 23)

### Splitting the data into train and test

In [67]:
df_processed['class'].value_counts()

negative                   3342
compensated_hypothyroid     187
primary_hypothyroid          91
secondary_hypothyroid         2
Name: class, dtype: int64

In [68]:
from sklearn.model_selection import train_test_split

In [69]:
X=df_processed.iloc[:,:-1]

In [70]:
X.shape

(3622, 22)

In [71]:
Y=df_processed['class']

In [72]:
Y.shape

(3622,)

In [73]:
# create training and testing variables
X_train, X_test, y_train, y_test = train_test_split(X,Y, test_size=0.25,random_state=45,stratify=df_processed['class'])

In [74]:
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(2716, 22) (2716,)
(906, 22) (906,)


### Logistic Regression

In [75]:
from sklearn.linear_model import LogisticRegression 
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

In [76]:
model = LogisticRegression()

In [77]:
model.fit(X_train,y_train)

C:\Users\VAISHAK\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\VAISHAK\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:460: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

In [78]:
predicted_class=model.predict(X_test)

In [79]:
accuracy = accuracy_score(y_test,predicted_class)*100

In [80]:
accuracy

92.27373068432672

In [81]:
print(classification_report(y_test,predicted_class))

                         precision    recall  f1-score   support

compensated_hypothyroid       0.00      0.00      0.00        47
               negative       0.92      1.00      0.96       836
    primary_hypothyroid       0.00      0.00      0.00        23

              micro avg       0.92      0.92      0.92       906
              macro avg       0.31      0.33      0.32       906
           weighted avg       0.85      0.92      0.89       906



C:\Users\VAISHAK\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1143: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


In [82]:
confusion_matrix(y_test, predicted_class)

array([[  0,  47,   0],
       [  0, 836,   0],
       [  0,  23,   0]], dtype=int64)

#### Not a good model as it is not predicting any other classes.

## Classification using neural network model

In [83]:
from sklearn.neural_network import MLPClassifier

In [84]:
nnmodel = MLPClassifier(hidden_layer_sizes=(100,60),random_state=45)

In [85]:
nnmodel.fit(X_train,y_train)

MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(100, 60), learning_rate='constant',
       learning_rate_init=0.001, max_iter=200, momentum=0.9,
       n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
       random_state=45, shuffle=True, solver='adam', tol=0.0001,
       validation_fraction=0.1, verbose=False, warm_start=False)

In [86]:
nnpredicted_class=nnmodel.predict(X_test)

In [87]:
accuracy_score(y_test,nnpredicted_class)*100

92.27373068432672

In [88]:
print(classification_report(y_test,nnpredicted_class))

                         precision    recall  f1-score   support

compensated_hypothyroid       0.75      0.06      0.12        47
               negative       0.93      1.00      0.96       836
    primary_hypothyroid       0.00      0.00      0.00        23

              micro avg       0.92      0.92      0.92       906
              macro avg       0.56      0.35      0.36       906
           weighted avg       0.89      0.92      0.89       906



In [89]:
confusion_matrix(y_test, nnpredicted_class)

array([[  3,  44,   0],
       [  1, 833,   2],
       [  0,  23,   0]], dtype=int64)

In [90]:
nnmodel = MLPClassifier(hidden_layer_sizes=(100,50),random_state=45)
nnmodel.fit(X_train,y_train)
nnpredicted_class=nnmodel.predict(X_test)
accuracy_score(y_test,nnpredicted_class)*100

92.27373068432672

In [91]:
nnmodel = MLPClassifier(hidden_layer_sizes=(100,70),random_state=45)
nnmodel.fit(X_train,y_train)
nnpredicted_class=nnmodel.predict(X_test)
accuracy_score(y_test,nnpredicted_class)*100

92.49448123620309

In [92]:
confusion_matrix(y_test,nnpredicted_class)

array([[  3,  44,   0],
       [  0, 835,   1],
       [  0,  23,   0]], dtype=int64)

In [93]:
nnmodel = MLPClassifier(hidden_layer_sizes=(100,60,30),random_state=45)
nnmodel.fit(X_train,y_train)
nnpredicted_class=nnmodel.predict(X_test)
accuracy_score(y_test,nnpredicted_class)*100

92.3841059602649

### SVM

In [94]:
from sklearn.svm import SVC
svclassifier = SVC(kernel='poly',degree=5, gamma='scale')
svclassifier.fit(X_train, y_train)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=5, gamma='scale', kernel='poly',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

In [95]:
svmpred = svclassifier.predict(X_test)

In [96]:
accuracy_score(y_test,svmpred)*100

92.27373068432672

### PCA

In [97]:
from sklearn.decomposition import PCA
pca = PCA()
x_pca = pca.fit_transform(X)
x_pca = pd.DataFrame(x_pca)
x_pca.head()

0         1         2         3         4         5         6   \
0  2.269829  0.187144 -0.427537 -0.024427 -0.043835 -0.012508 -0.064432   
1 -0.787513  0.646122 -0.103938 -0.832324  0.368139 -0.258670  0.010176   
2 -0.724092 -0.014514  0.705770  0.623391 -0.395664 -0.094843 -0.011154   
3 -0.851538  0.584966 -0.230882 -0.368465  1.224269 -0.193461 -0.072433   
4  0.309670 -0.167105 -0.335132 -0.165555 -0.083285 -0.090394 -0.063295   

         7         8         9   ...        12        13        14        15  \
0  0.029811 -0.048059 -0.225492  ... -0.021457  0.013037  0.021976 -0.066424   
1 -0.538815  0.049113 -0.121758  ...  0.285108  0.033753 -0.035198 -0.152617   
2 -0.297266 -0.083965 -0.035129  ... -0.005676 -0.018571 -0.032215 -0.005426   
3 -0.280498 -0.255815  0.050178  ...  0.330276  0.001392 -0.042897 -0.132930   
4  0.082010 -0.116824 -0.018391  ... -0.011535 -0.018323 -0.006910 -0.004324   

         16        17        18        19        20        21  
0  0.021023  0.012397 -0.006237 -0.027622 -0.000248 -0.000606  
1 -0.108650 -0.013645 -0.015270 -0.002434 -0.000603 -0.000573  
2 -0.014483 -0.015713 -0.017393 -0.003384 -0.000096 -0.000481  
3 -0.138430  0.001981  0.005034  0.000491  0.000300  0.001098  
4  0.001369 -0.002446 -0.003218 -0.006378 -0.000310  0.000320  

[5 rows x 22 columns]

In [98]:
explained_variance = pca.explained_variance_ratio_
explained_variance

array([5.18258987e-01, 1.18843619e-01, 9.17934367e-02, 5.78021324e-02,
       4.12576568e-02, 2.62065862e-02, 2.56006158e-02, 1.91047147e-02,
       1.89108012e-02, 1.56782207e-02, 1.52232217e-02, 1.08675662e-02,
       6.60525740e-03, 6.13865542e-03, 6.05397564e-03, 5.62062390e-03,
       5.03968419e-03, 4.75561195e-03, 3.92287897e-03, 2.08236175e-03,
       1.17732351e-04, 1.15659823e-04])

In [99]:
len(explained_variance)

22

In [100]:
s=0
for i in range(len(explained_variance)):
    if s < 0.96:
        s+=explained_variance[i]
    else:
        print(explained_variance[i])
        break

0.006138655416832092


Excluding last 9 columns (till the above value of variance) to take only the relevant features for the model.

In [101]:
X.shape

(3622, 22)

In [102]:
cols=[13,14,15,16,17,18,19,20,21]
X=X.drop(X.columns[cols],axis=1)

In [103]:
X.head()

sex  on_thyroxine  query_on_thyroxine  on_antithyroid_medication  sick  \
0    0             0                   0                          0     0   
1    0             0                   0                          0     0   
2    1             0                   0                          0     0   
3    0             1                   0                          0     0   
4    0             0                   0                          0     0   

   pregnant  thyroid_surgery  I131_treatment  query_hypothyroid  \
0         0                0               0                  0   
1         0                0               0                  0   
2         0                0               0                  0   
3         0                0               0                  0   
4         0                0               0                  0   

   query_hyperthyroid  lithium  goitre  tumor  
0                   0        0       0      0  
1                   0        0       0      0  
2                   0        0       0      0  
3                   0        0       0      0  
4                   0        0       0      0

In [104]:
X_train, X_test, y_train, y_test = train_test_split(X,Y, test_size=0.25,random_state=45,stratify=df_processed['class'])

In [105]:
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(2716, 13) (2716,)
(906, 13) (906,)


In [106]:
model = LogisticRegression(solver='lbfgs',multi_class='multinomial')

In [107]:
model.fit(X_train,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='multinomial',
          n_jobs=None, penalty='l2', random_state=None, solver='lbfgs',
          tol=0.0001, verbose=0, warm_start=False)

In [108]:
predicted_class=model.predict(X_test)

In [109]:
accuracy_score(y_test,predicted_class)*100

92.27373068432672

### Random forest

In [110]:
from sklearn.ensemble import RandomForestClassifier

In [111]:
rdfmodel = RandomForestClassifier(n_estimators=200)

In [112]:
rdfmodel.fit(X_train,y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=200, n_jobs=None,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

In [113]:
rdfpredicted_class=rdfmodel.predict(X_test)

In [114]:
accuracy_score(y_test,rdfpredicted_class)*100

92.27373068432672

Accuracy does not seem to change much. 

#### Best model is the neural network with 2 hidden layers having 100 and 70 nodes. The inaccuracy to predict other classes can be connected to the imbalance in the data which is understood by the df_processed.value_counts(). But this model has been able to predict other classes also.